### ISSUE TITLE
`ApiextensionsV1beta1Api()` not part of kubernetes 3.0.0

### ISSUE THREAD

**Issue by sebgoa** [kubernetes-incubator/client-python#381]
I am trying to test CRD and I re-installed the kubernetes module, version3.0.0 . Using ipyton 2.7

```
)$ sudo pip install --upgrade kubernetes
The directory '/Users/sebgoa/Library/Caches/pip/http' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
The directory '/Users/sebgoa/Library/Caches/pip' or its parent directory is not owned by the current user and caching wheels has been disabled. check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
Collecting kubernetes
  Downloading kubernetes-3.0.0-py2.py3-none-any.whl (815kB)
    100% |████████████████████████████████| 819kB 1.0MB/s 
Requirement already up-to-date: setuptools>=21.0.0 in /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from kubernetes)
Requirement already up-to-date: requests in /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from kubernetes)
Requirement already up-to-date: urllib3!=1.21,>=1.19.1 in /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from kubernetes)
Requirement already up-to-date: google-auth>=1.0.1 in /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from kubernetes)
Requirement already up-to-date: certifi>=14.05.14 in /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from kubernetes)
Requirement already up-to-date: pyyaml>=3.12 in /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from kubernetes)
Requirement already up-to-date: python-dateutil>=2.5.3 in /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from kubernetes)
Requirement already up-to-date: websocket-client<=0.40.0,>=0.32.0 in /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from kubernetes)
Requirement already up-to-date: six>=1.9.0 in /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from kubernetes)
Requirement already up-to-date: ipaddress>=1.0.17 in /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from kubernetes)
Requirement already up-to-date: idna<2.7,>=2.5 in /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from requests->kubernetes)
Requirement already up-to-date: chardet<3.1.0,>=3.0.2 in /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from requests->kubernetes)
Requirement already up-to-date: cachetools>=2.0.0 in /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from google-auth>=1.0.1->kubernetes)
Requirement already up-to-date: pyasn1-modules>=0.0.5 in /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from google-auth>=1.0.1->kubernetes)
Requirement already up-to-date: pyasn1>=0.1.7 in /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from google-auth>=1.0.1->kubernetes)
Requirement already up-to-date: rsa>=3.1.4 in /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from google-auth>=1.0.1->kubernetes)
Requirement already up-to-date: backports.ssl_match_hostname in /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages (from websocket-client<=0.40.0,>=0.32.0->kubernetes)
Installing collected packages: kubernetes
Successfully installed kubernetes-3.0.0
sebimac:kkcd (master)$ ipython ./testcrd.py 
default
kube-public
kube-system
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/Users/sebgoa/Desktop/kkcd/testcrd.py in <module>()
      8      print ns.metadata.name
      9 
---> 10 crd = client.ApiextensionsV1beta1Api()
     11 

AttributeError: 'module' object has no attribute 'ApiextensionsV1beta1Api'
```

**Comment by sebgoa** [comment]
It is in HEAD but not in the available module on Pypi.

**Comment by eyeofthefrog** [comment]
You can install 4.0.0a1 via pip using this command:
```
pip install kubernetes==4.0.0a1 --upgrade
```

That'll give you support for CRD. I agree it should be supported in 3.0.0, this is just a work around so you can continue with development.

**Comment by sebgoa** [comment]
thanks

**Comment by eyeofthefrog** [comment]
Why is this closed? We need this in 3.x.

**Comment by mbohlool** [comment]
if ApiextensionsV1beta1 is a new API in 1.8 then it will be supported in 4.x clients. There is a compatibility matrix in the main readme file for reference.

**Comment by sebgoa** [comment]
I opened it, I am happy with the answer so I closed it.

**Comment by eyeofthefrog** [comment]
@mbohlool This has been in since 1.7. Also, the compatibility matrix says that the 3.0 python client supports kubernetes 1.8. 

@sebgoa The original problem you stated still exists. You personally have a work-around, but the behavior of the 3.0 python client is still incorrect. This is not the right way to handle issues. Please let me know if you're unwilling to reopen this issue and I'll open another.

**Comment by sebgoa** [comment]
the problem is more in the compatibility matrix I think than in the content of the client.

**Comment by sebgoa** [comment]
and fwiw, don't ask me to close this issue in the future

**Comment by eyeofthefrog** [comment]
I disagree, this is functionality that is in kubernetes 1.7....unless you're saying that the 3.0 python client shouldn't support 1.7 either.

**Comment by rahulinux** [comment]
+1 should be available in 3.0.0

**Comment by mbohlool** [comment]
The compatibility matrix has a bug in it I will fix (3.0 supports kubernetes 1.7) and in 1.7 spec https://raw.githubusercontent.com/kubernetes/kubernetes/release-1.7/api/openapi-spec/swagger.json there is no api extension. If you believe the functionality is available in 1.7, that should be a bug in 1.7 that it does not expose api extensions in OpenAPI spec. I would follow up with an issue there, but the best course of action I think, is to upgrade to 4.0.0. Is there anything blocking you guys from the upgrade?

**Comment by eyeofthefrog** [comment]
We can move to 4.0.0. When the issue was open, the python client 4.0.0 was still alpha. Even so, we moved to it in order to get support for the ApiextensionsV1beta1Api on our 1.7.5 kubernetes cluster. So yes, this API should be exposed for the 1.7 release in the OpenAPI spec. Can you open a bug or point me to the correct project so I can open a bug? Thanks

From a user's perspective, it seems like a client issue which is why the conversation is happening here. Is it correct for me to assume there's a code generation tool that creates the python APIs from the swagger spec, which is why the API exists in the 4.0.0 client?

**Comment by mbohlool** [comment]
Yes, there is a code generator [here](https://github.com/kubernetes-client/gen). It is in 4.0 client because kubernetes 1.8 exposed the API in their spec and we re-generated the client using new Spec. The project you need to create the issue on is the main kubernetes repo to get the spec fixed for 1.7 but honestly, I don't think it will get much attention as we are releasing 1.9 kubernetes this week and the promise of support for old kubernetes version is only last version and only critical fixes. Here people can upgrade to 4.0 client and get what they need. If we (as client python developers) see a good reason people don't want to upgrade to 4.0 and still need this, we can discuss some solutions here, but I don't see it critical. I will close this, please let me know if you need more help here and we can re-open it.

**Comment by eyeofthefrog** [comment]
Since we have 4.0.0 out of alpha, I agree we can "forget the past" as it were. I do have another bug with the 4.0.0 client and the 1.8.0 server concerning this same API which I'll open shortly. Thanks for your help.
